### PR TITLE
Create array-permutations package

### PR DESCRIPTION
**Pull request by EvandroLG**
This function returns all permutations of the length N of the elements of a given Array.
I also include a small refactoring in `groupBy` function.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1618,6 +1619,20 @@ const getRoot = partial(Math.pow, 64);
...
+:icecream:[`Try It`](https://anguscroll.com/just/just-permutations)
+
+`npm install just-permutations`
+
+```js
+import permutations from "just-permutations";
+
+permutations([1, 2, 3]); // [[1, 2, 3], [2, 1, 3], [2, 3, 1], [1, 3, 2], [3, 1, 2], [3, 2, 1]]
+permutations([]); // []
+random(); // throws
```
Comment (angus-c) [inline README.md:1633]: `random()` -> `permutations()`

File: test/array-permutations/index.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    permutations([1, 2, 3]),
+    [[1, 2, 3], [2, 1, 3], [2, 3, 1], [1, 3, 2], [3, 1, 2], [3, 2, 1]]
+  );
+
+  t.deepEqual(permutations([]), []);
+
+  t.end();
+});
+
+test('invalid', function(t) {
```
Comment (angus-c) [inline test/array-permutations/index.js:22]: Also add test for `permutations()`

**Review by angus-c — CHANGES_REQUESTED** [review]
Many thanks for this! Can you also patch bump version in array-group-by `package.json`

File: packages/array-permutations/index.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  if (arr.length === 1) {
+    return [arr];
+  }
+
+  const output = [];
+  const partialPermutations = permutations(arr.slice(1));
+  const first = arr[0];
+
+  for (let i = 0; i < partialPermutations.length; i++) {
```
[the remark this answers is not in the document]
Comment (EvandroLG) [reply packages/array-permutations/index.js:25]: As far as I know the `length` is cached internally in modern engines.
That said, do you think it is still relevant to cache the `length`? If so, could I ask you why? :)

[the remark this answers is not in the document]
Comment (angus-c) [reply packages/array-permutations/index.js:25]: Well we also support older engines. In normal code I would not go so far as to cache length, but these utilities could be part of very hot code (called very often).

[the remark this answers is not in the document]
Comment (EvandroLG) [reply packages/array-permutations/index.js]: Okay. I also replaced `const` and `let` with `var`.

**Comment by angus-c** [comment]
Landed: https://anguscroll.com/just/just-permutations
Thanks again!

**Pull request merged**